interactive/submit.py

```diff
@@ -0,0 +1,240 @@
+        requested_actions=["run_all"],
+    )
+
+    analysis_request.job_request = job_request
+    analysis_request.save()
+
+    # TODO: notify someone about output checking?
+    # waiting to find out what this should be
+    # issue_url = create_issue(analysis_request.pk, job_server_url=url)
+    # notify_analysis_request_submitted(analysis_request, issue_url)
+
+    # TODO: notify the user?
+    # send_analysis_request_confirmation_email(
+    #     analysis_request.user.email, analysis_request
+    # )
+    return analysis_request
+
+
+# MOVE TO EXTERNAL SERVICE
+
+
+def clean_working_tree(path):
+    """Remove all files (except .git)"""
+    for f in path.glob("**/*"):
+        relative = f.relative_to(path)
+
+        if str(relative) == ".git" or str(relative).startswith(".git/"):
+            continue
+
+        print(f, relative)
+
+        f.unlink() if f.is_file() else shutil.rmtree(f)
+
+
+def commit_and_push(working_dir, analysis, force=False):
+    force_args = ["--force"] if force else []
+
+    git("add", ".", cwd=working_dir)
+
+    second_codelist = ""
+    if analysis.codelist_2:
+        second_codelist = f" and codelist {analysis.codelist_2.slug}"
+    msg = (
+        f"Codelist {analysis.codelist_1.slug}{second_codelist} ({analysis.identifier})"
+    )
+
+    git(
+        # -c arguments are instead of having to having to maintain stateful git config
+        "-c",
+        "user.email=interactive@opensafely.org",
+        "-c",
+        "user.name=OpenSAFELY Interactive",
+        "commit",
+        "--author",
+        f"{analysis.created_by} <{analysis.created_by}>",
+        "-m",
+        msg,
+        cwd=working_dir,
+    )
+    ps = git("rev-parse", "HEAD", capture_output=True, cwd=working_dir)
+    commit_sha = ps.stdout.strip()
+
+    # this is an super important step, makes it much easier to track commits
+    git("tag", analysis.identifier, *force_args, cwd=working_dir)
+
+    # push to main. Note: we technically wouldn't need this from a pure git
+    # pov, as a tag would be enough, but job-runner explicitly checks that
+    # a commit is on the branch history, for security reasons
+    git("push", "origin", "main", "--force-with-lease", cwd=working_dir)
+
+    # push the tag once we know the main push has succeeded
+    git("push", "origin", analysis.identifier, *force_args, cwd=working_dir)
+    return commit_sha
+
+
+def create_commit(
+    analysis,
+    template_repo,
+    force=False,
+    get_opencodelists_api=_get_opencodelists_api,
+):
+    if not force:
+        # check this commit does not already exist
+        raise_if_commit_exists(analysis.repo, analysis.identifier)
+
+    # 1. create tempdir with AR.pk suffix
+    suffix = f"template-{analysis.identifier}"
+    with tempfile.TemporaryDirectory(suffix=suffix) as working_dir:
+        working_dir = Path(working_dir)
+
+        # 2. clone the given analysis code template repo to tempdir
+        git("clone", template_repo, working_dir)
```

Comment:
worth doing a shallow clone here, for speed in tests if nothing else?

Reply:
Done for both clones as I don't think we care about history for either repo.